src/bittybuzz/src/BittyBuzzMathFunctions.cpp

```diff
@@ -0,0 +1,311 @@
+    if (bbztype_isfloat(*o)) {
+        arg = bbzfloat_tofloat(o->f.value);
+    } else if (bbztype_isint(*o)) {
+        arg = o->i.value;
+    } else {
+        bbzvm_seterror(BBZVM_ERROR_MATH);
+        return;
+    }
+
+    bbzvm_pushf(bbzfloat_tofloat(Math::tan(arg)));
+    bbzvm_ret1();
+}
+void BittyBuzzMathFunctions::bbzmath_asin() {
+    bbzvm_assert_lnum(1); // NOLINT
+
+    float arg = 0;
+    bbzobj_t* o = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    if (bbztype_isfloat(*o)) {
+        arg = bbzfloat_tofloat(o->f.value);
+    } else if (bbztype_isint(*o)) {
+        arg = o->i.value;
+    } else {
+        bbzvm_seterror(BBZVM_ERROR_MATH);
+        return;
+    }
+
+    bbzvm_pushf(bbzfloat_tofloat(Math::asin(arg)));
+    bbzvm_ret1();
+}
+
+void BittyBuzzMathFunctions::bbzmath_acos() {
+    bbzvm_assert_lnum(1); // NOLINT
+
+    float arg = 0;
+    bbzobj_t* o = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    if (bbztype_isfloat(*o)) {
+        arg = bbzfloat_tofloat(o->f.value);
+    } else if (bbztype_isint(*o)) {
+        arg = o->i.value;
+    } else {
+        bbzvm_seterror(BBZVM_ERROR_MATH);
+        return;
+    }
+
+    bbzvm_pushf(bbzfloat_tofloat(Math::acos(arg)));
+    bbzvm_ret1();
+}
+
+void BittyBuzzMathFunctions::bbzmath_atan() {
+    bbzvm_assert_lnum(2); // NOLINT
+
+    float y = 0;
+    bbzobj_t* a = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    if (bbztype_isfloat(*a)) {
+        y = bbzfloat_tofloat(a->f.value);
+    } else if (bbztype_isint(*a)) {
+        y = a->i.value;
+    } else {
+        bbzvm_seterror(BBZVM_ERROR_MATH);
+        return;
+    }
+
+    float x = 0;
+    bbzobj_t* b = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    if (bbztype_isfloat(*b)) {
+        x = bbzfloat_tofloat(b->f.value);
+    } else if (bbztype_isint(*b)) {
+        x = b->i.value;
+    } else {
+        bbzvm_seterror(BBZVM_ERROR_MATH);
+        return;
+    }
+
+    bbzvm_pushf(bbzfloat_tofloat(Math::atan2(y, x)));
+    bbzvm_ret1();
+}
+
+void BittyBuzzMathFunctions::bbzmath_min() {
+    bbzvm_assert_lnum(2); // NOLINT
+    /* Get arguments */
+    bbzobj_t* a = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    bbzobj_t* b = bbzheap_obj_at(bbzvm_locals_at(2)); // NOLINT
+
+    /* Compare them and return the smaller one */
+    int cmp = bbztype_cmp(a, b);
+    bbzvm_lload(cmp >= 0 ? 1 : 2);
+    bbzvm_ret1();
+}
+
+void BittyBuzzMathFunctions::bbzmath_max() {
+    bbzvm_assert_lnum(2); // NOLINT
+    /* Get arguments */
+    bbzobj_t* a = bbzheap_obj_at(bbzvm_locals_at(1)); // NOLINT
+    bbzobj_t* b = bbzheap_obj_at(bbzvm_locals_at(2)); // NOLINT
+
+    /* Compare them and return the bigger one */
+    int cmp = bbztype_cmp(a, b);
+    bbzvm_lload(cmp >= 0 ? 1 : 2);
+    bbzvm_ret1();
+}
+
+void BittyBuzzMathFunctions::bbzmath_rng_uniform() {
+    bbzvm_assert_lnum(0); // NOLINT
+    uint32_t randomNum = BittyBuzzSystem::g_bsp->generateRandomNumber();
```

Comment:
Could it be changed to get the reference from the container? would be safer than linking to an external pointer with no null check...

Reply:
No it can't since the namespace cannot initialize a reference, it has no constructor. We need to inject it via the bbvm so it's possible to mock. Also this pointer should never be null since this is only called after the bbvm is initialized, and the bbvm takes a reference at it's constructor to initialize this pointer. In other words. the BittyBuzzSystem is initialized by the BittyBuzzVm.